### PR TITLE
fix(s13a.13): self-ref look-back no-prior short-circuit — Phase 6 #3f

### LINKED ISSUE #76: S13a.13:  must resolve to "foo" when a has no prior value (spec L841)
## Description

Per HOCON spec L841, the expression `a = ${?a}foo` must resolve to `"foo"` because `a` has no prior assigned value, so the self-referencing optional substitution `${?a}` is undefined and contributes an empty string.

## Current behaviour

```hocon
a = ${?a}foo
```

Resolves to `"foofoo"`. The resolver appears to see the already-computed value `"foo"` for `a` when evaluating `${?a}`, creating a cycle-like double-evaluation.

## Expected behaviour

`a` must resolve to `"foo"`. With no prior definition of `a`, `${?a}` is undefined → empty string → concatenated with `"foo"` → `"foo"`.

## Spec reference

HOCON.md L841 (Self-Referential Substitutions): `a = ${?a}foo` produces `"foo"` when `a` has no prior value.

## Pinning test

`tests/integration_test.rs`: `s13a_13_optional_self_ref_concat_with_no_prior_pin` (asserts current `"foofoo"` behaviour)  
`tests/integration_test.rs`: `s13a_13_optional_self_ref_concat_with_no_prior_spec` (`#[ignore]` — asserts correct `"foo"`)

## Phase

Identified in Phase 3 spec-compliance audit.

### PR DESCRIPTION
**Pull request by y1o1**
## Summary

Fix S13a.13: `a = ${?a}foo` with no prior `a` now resolves to `"foo"` (per HOCON.md L841), was `"foofoo"`.

Closes #76.

## Mechanism

When the resolver encounters a self-referential substitution `${?a}` (or `${a}`) inside the value being assigned to path `a`:
1. If a prior assignment to `a` exists → look-back returns the prior value (unchanged behavior).
2. **NEW**: If no prior exists → short-circuit. Optional `${?a}` → `Ok(None)` (concat-fold omits it); required `${a}` → `ResolveError`.

Identification: a substitution is "self-referential" when **the currently-being-assigned field path equals the substitution path**. Tracked via a `resolving_field_path: Vec<String>` stack on the resolver (push leaf key on entry to each `for (key, val)` iteration in `resolve_res_obj`, pop on exit).

Additional fix (sr09/sr10 nested support): when prior is found and the substitution path is multi-segment (e.g. `foo.a`), navigate into the prior root object via `lookup_path(prior_obj, &s.segments[1..])` to find the nested leaf value.

## Why not the spec's path-equality alone?

The original spec ★1 decision #1 specified path-equality preservation for self-ref detection. Round-2 multi-agent-review surfaced a false-positive: external lookups like `a = ${?a}foo; b = ${a}` would error because `b`'s `${a}` lookup found `a`'s concat-with-self-ref and triggered the short-circuit (wrongly).

All 3 impls (ts/rs/go) independently tightened the mechanism. ts uses node-membership-in-iterating-set; rs uses this path stack + `is_owner` guard; go uses AST-node pointer-identity. **All three are strictly narrower than path-equality** (correctly exclude external lookups).

Spec amendment deferred to a follow-up xx.hocon PR. See [xx.hocon#27](https://github.com/o3co/xx.hocon/issues/27) for the cross-impl tracking issue.

## Test plan

- [x] Conformance fixtures sr01–sr11 (xx.hocon `testdata/hocon/self-ref-lookback/`) all pass
- [x] Integration tests for 11+ edge cases (no-prior, with-prior, required, array, nested, mutual-ref, external-ref regression guard)
- [x] Round-2 regression test: `a = ${?a}foo; b = ${a}` → both `"foo"`
- [x] Existing `s13a_13_optional_self_ref_concat_with_no_prior_pin` test removed; `s13a_13_..._spec` un-ignored
- [x] env_var_list_test.rs ev08 tripwire comments removed
- [x] `cargo test` green (~524 tests)
- [x] `cargo clippy --lib -- -D warnings` clean (pre-existing test-target clippy issues are not introduced by this PR)
- [x] `cargo fmt --check` clean
- [x] No public API changes (`SubstitutionResolver` is `pub(crate)`; no new public types/enums)

## Pre-existing bugs surfaced (out of scope, see xx.hocon#27)

Round 2 review discovered 4 pre-existing resolver bugs beyond S13a.13 strict:
- Nested external ref crash (`foo.a = ${?foo.a}bar; foo.b = ${foo.a}`)
- Cache pollution on prior-with-external-ref
- Same-field double-self-ref crash
- Order-dependent ref-before-self-ref

These are filed as cross-impl follow-ups; not addressed in this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/resolver/substitution_resolver.rs [hunk 1]
```diff
@@ -155,27 +179,76 @@ impl<'a> SubstitutionResolver<'a> {
...
+                        _ => false,
+                    };
                 if is_self_ref {
                     let root_seg = s.segments.first().map(|s| s.text.as_str()).unwrap_or("");
-                    let prior = scope
+                    let prior_root = scope
                         .prior_values
                         .get(root_seg)
                         .or_else(|| self.root.prior_values.get(root_seg))
                         .cloned();
```
[the remark this answers is not in the document]
Comment (y1o1) [reply src/resolver/substitution_resolver.rs:214]: Fixed in 3c264b0 — verified empirically the test was RED. The object-literal form `foo { a = "x"; a = ${?foo.a}bar }` stores leaf prior in foo's `scope.prior_values["a"]`, not in `root.prior_values["foo"]`. Added a leaf-segment fallback after the root-segment lookup fails. Regression test `s13a_13_nested_self_ref_object_literal_form` added as guard.

File: tests/self_ref_lookback_test.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    );
+
+    let env: HashMap<String, String> = HashMap::new();
+    let result = hocon::parse_file_with_env(&fp, &env);
+
+    if has_error {
+        assert!(
+            result.is_err(),
+            "self-ref-lookback {}: expected parse/resolve error but got Ok (fixture: {})",
+            stem,
```
[the remark this answers is not in the document]
Comment (y1o1) [reply tests/self_ref_lookback_test.rs:148]: Fixed in 3c264b0 — added explicit `fp.exists()` assertion in `run_fixture` before the sidecar check. Missing or misspelled fixture paths now panic with a clear diagnostic ("fixture missing — run \`make testdata\`") instead of silently passing as expected-error.

File: src/resolver/substitution_resolver.rs [hunk 3]
```diff
@@ -155,27 +179,76 @@ impl<'a> SubstitutionResolver<'a> {
...
                         }
-                        return Ok(result);
+                        // Prior root exists but nested path not found — fall through to
+                        // no-prior short-circuit below.
+                    }
+                    // Spec L841: no prior + self-ref → optional yields undefined; required errors.
+                    if s.optional {
+                        // Cache the undefined result for idempotency (spec Q2).
+                        // Return None (undefined) — the concat-layer optional-omission rule
+                        // (Phase 6 #3b) will omit this from the fold input.
```
[the remark this answers is not in the document]
Comment (y1o1) [reply src/resolver/substitution_resolver.rs]: Fixed in 3c264b0 — comment updated to reflect reality. Cache stores `HoconValue`, not `Option<HoconValue>`; undefined is encoded by absence (no explicit cache entry needed; re-resolving the same self-ref deterministically hits the same short-circuit). Convergent with Codex r1 M1 finding.
